### PR TITLE
Add grok CLI, Qwen Code, Ori and Cline, and stop our Claude hooks running inert inside grok

### PR DESCRIPTION
**Pull request by chhhee10**
> **Draft — not for merge yet.** Opened to get the work reviewable; a few things below are deliberately unfinished.

Adds **grok CLI** (xAI) and **Qwen Code** as the 13th and 14th integrations — live-hook enforcement *and* audit, user + project scope — and fixes a silent non-enforcement bug that affects people who have installed neither.

Every contract claim here was verified against the running CLI with a recorder hook on every event, plus deny and stop-gate probes. Where a vendor's docs and the wire disagreed, the wire won and the disagreement is recorded in `CLAUDE.md`.

## The fix that matters without either integration

grok's hook discovery scans `~/.claude/settings.json`, `~/.claude/settings.local.json` and `<cwd>/.claude/settings.json` **by default**. That last one is exactly what `policies --install --cli claude --scope project` writes — so on any machine with both tools, grok was already executing our hooks, passing `--cli claude` while piping its own camelCase payload.

`tool_name` and `tool_input` arrived `undefined`, so every builtin that matches a tool name or reads a command or path (`block-sudo`, `block-env-files`, `block-secrets-write`, `block-force-push`) saw nothing and allowed. A deny would not have landed anyway: grok ignores Claude's `hookSpecificOutput` shape — proven by A/B on one live hook, where Claude's shape let the command run and grok's blocked it.

Installed, running, costing latency, enforcing nothing — worse than no coverage, because the install reports success. Same class as the Copilot input-key drift and the Hermes `subagent_stop` row.

The handler now detects grok's envelope **from the payload shape alone** (never an env var, so a real Claude event cannot be misread) and routes it onto grok's contract — tool maps and response shape both. Real Claude payloads are untouched.

## Three things only probing found

- grok's `read_file` delivers the path as `target_file` (and `list_dir` as `target_directory`). Without the input map, a live `.env` read walks past `block-env-files`.
- grok fires `Stop` **twice**, the second at shutdown with its decision parsed and discarded. Blocking there records enforcement that can never happen, so the gate keys on `reason === "end_turn"`.
- grok discovers project hooks **only inside a git repo**. A trusted non-git directory holding a valid config logs `project_sources=0` and never fires — silently.

qwen needed none of that: its payload is pure Claude snake_case and all six tools deliver canonical keys. Its traps are documented rather than worked around — `stop_hook_active` is `true` on the *first* Stop fire (unusable as a loop guard), and `UserPromptSubmit` fires once per model invocation, seven times in one observed turn.

## Coverage

Both reach **35 of 40 builtin policies enforcing** — day-one parity with claude/cursor/devin/factory/antigravity/openclaw. The 5 that don't are the `sanitize-*` family on `PostToolUse`, which only codex and copilot can block; that gap is fleet-wide, not specific to these two.

Event surface is each CLI's full useful set: grok **14 of its own 14**, qwen **19**. Verified before implementing — grok silently skips event keys it does not recognise, and qwen turned out to dispatch three events its own docs never list (`InstructionsLoaded`, `UserPromptExpansion`, `PostToolBatch`). Three are deliberately left out with measured reasons: `MessageDisplay` fires per streaming chunk, `PostToolBatch` measured +76% hook invocations while duplicating `PostToolUse`, and `SessionDelete` has no canonical equivalent.

qwen's `TodoCreated`/`TodoCompleted` map onto `TaskCreated`/`TaskCompleted` and are a **real veto point** — they run in a `validation` phase where a block prevents the write, verified live by blocking a todo that planned to skip the tests.

## Also in here

- **Daemon collection.** Two new `fpai-collect` sources so their sessions reach the cloud like every other CLI's. qwen follows the Factory model; grok follows the **cursor** model, because its transcript carries no timestamps at all — events are stamped from file mtime plus byte offset, which keeps time approximately right while staying a pure function of the inputs, as the content-hash dedup requires.
- **Dashboard.** Both in the CLI filter and the projects list, with badges off the status palette (status hues carry health meaning and must not be spent on identity).
- **`policies --install --cli grok` was rejected outright** — a third hardcoded CLI list in `bin/failproofai.mjs`, in three copies, which the unit suite could not see. Now one list plus a tripwire test. Found only by packing and installing the tarball.
- **grok projects 404'd** from the projects list: its percent-encoded folder name became the URL slug and re-encoded to `%252F…`. Found by clicking the link, not by reading the page.
- **Three `fp-reset` tests** that passed or failed depending on whether the developer happened to have a daemon installed.

## Verified end to end

Against both live CLIs with the packed tarball installed globally, and against a real `failproofaid` systemd service: `block-sudo` and `block-env-files` blocked on each, all response shapes correct, and 197 grok + 207 qwen events delivered to an ingest endpoint — transcripts *and* hook activity, including sessions started after the daemon was already running.

## Why it is a draft

- Some capability rows are deliberately `observe` rather than `block`: grok `SubagentStop` (advertised in its own `blockingEvents`, never exercised) and qwen `UserPromptSubmit`/`PostToolUse`/`SubagentStop`. They are probably blocking; nobody has proven it, so nothing claims it.
- **Whether `PreToolUse` fires for tools called inside a subagent is unverified on both.** If it does not, that is an unguarded execution path that currently reads as covered — worth settling before this merges.
- qwen may support a `PostToolUse` block, which would take it to 40/40 and flip the five `sanitize-*` rows. One probe would answer it.
- The 14 translated READMEs still show twelve CLIs; those regenerate from `scripts/translate-docs`.

Full gate green on the rebased tree: **3864 unit**, **320 e2e**, **393 rust**, `tsc`, `clippy`, `fmt` and lint clean, and `bun run build` passes.

---

## Added since: Ori (OpenRouter), the 15th integration

`ori` is two products behind one binary, and only the second needed anything built.

As a **launcher** (`ori claude`, `ori codex`, `ori grok`, `ori opencode`, …) it injects credentials and nothing else — no `HOME` or config-dir redirect — and its injections **merge** rather than replace. Proven, not assumed: a project `SessionStart` hook fired identically with and without ori's exact `--settings` blob, and against `opencode debug config` the `plugin` array survived `OPENCODE_CONFIG_CONTENT` intact. **The hooks already installed for those CLIs keep enforcing under `ori <agent>`, so nothing ships for that path.**

What ships gates ori's **own** agent (bare `ori` / `ori code`) through its published approval extension points — a feature directory generated at `~/.ori/global/features/failproofai/`, auto-discovered with no config file to register it in, like Goose's plugin dir. User scope only, because bare `ori` boots the global workspace.

**Read this before demoing.** ori's approval mode defaults to `self-drive`, and in that mode the dynamic approval points are never called, so policies never run. Isolated three ways: no callback under self-drive; still none after claiming `approval-policy` with `defaultAction:"ask"`; but `defaultAction:"reject"` *did* block every call. No config key or env var changes that default, so **`--approvals manual` is a documented requirement**, and we claim `approval-policy` with an inert `ask` rather than a blanket `reject` so installing cannot brick a self-drive session.

In exchange ori is the **only integration that fails closed for free** — both dynamic points declare `failureBehavior: "deny"`, so a failproofai fault blocks the call. Goose and OpenClaw fail open today.

`edit` is the sharp edge: it carries the whole change as one apply_patch blob with **no path argument**, so `file_path` was absent and every path builtin would have silently no-opped on edits. `oriPatchFilePaths()` recovers it; the multi-file limitation is asserted in a test rather than left to be discovered.

Audit reads the transcript from `~/.ori/global/.ori/state.sqlite`, whose misnamed `ori_agent_loop_history.prompt` column holds the entire serialized conversation. Three plausible stores are dead ends and are recorded so nobody re-derives them.

**Verified live** against ori 0.12.0+68f9a36 driving `nvidia/nemotron-3.5-lightning:free`: the shipped install path blocks a real `.env` read via the `block-env-files` builtin (2 `tool.started`, 2 `tool.failed`, zero succeeded — the secret never reached the model), and uninstall removes the feature cleanly.

### Two things the rebase surfaced, fixed here

- **grok and Qwen were missing from `docs/reference/harnesses.mdx` entirely.** That file and the `copy-counts` test guarding it both reached `main` *after* this branch was cut, so the gap only becomes visible on a rebase. Both now appear in the scopes, config-path, enforcement, conditions and probed-version tables, and every count across the README, docs and package description moves to **15 harnesses / 15 pre-tool gates / 10 stop gates**, derived from source rather than restated.
- **`bin/failproofai.mjs` had three hand-written `--cli` lists.** Two already read a shared `INSTALLABLE_CLIS`; the third was still a literal — the same shape that made `--install --cli grok` fail. All three now read the array, and `cli-flag-coverage.test.ts` fails if a fourth literal appears or the array drifts from `INTEGRATION_TYPES`. It also covers the `--hook` guard, whose fallback to `"claude"` for an unlisted CLI is **silent**: the hook runs, evaluates, records activity, and answers in Claude's wire shape to a CLI that cannot parse it.

> The ori logo in the README is a plain typographic placeholder — OpenRouter publishes no standalone ori logo asset, and every other logo in that grid is a real brand file. Worth swapping if an official one appears.

---

## Added since: Cline, the 16th integration

Cline is the only integration whose config is a **directory of event-named files**. Three rules govern it and **all three fail silently when broken**: the filename IS the event (ten names, case-insensitive), the extension must be in cline's allowlist, and the exit code is ignored — the verdict is the single JSON object on stdout. `--hooks-dir` is a **dead flag** in v3.0.60: it writes `CLINE_HOOKS_DIR`, which appears exactly once in the shipped binary (that write) and is read by nothing. An earlier probe concluded cline had no reachable hook surface; it had the right directory and the wrong filenames, and cline skips a non-matching name without a log line.

**Every cline tool is BATCH-shaped, and that is the whole integration.** `run_commands{commands[]}`, `read_files{files[{path}]}`, `search_codebase{queries[]}` and one multi-file `apply_patch{input}` blob — against builtins that read the scalars `tool_input.command` / `.file_path` / `.pattern`. A key rename would leave every one of them reading `undefined` and **allowing silently**, the inert-hook failure this repo has already shipped twice. So there is deliberately no `CLINE_TOOL_INPUT_MAP`: `batch-expand.ts` expands one call into N canonical scalar inputs and `batch-fanout.ts` runs the policy set once per element, lowest offending index winning and short-circuiting the rest.

**Joining the array was designed, judged and rejected on evidence.** `SECRET_FILE_RE` is `/\.(?:pem|key)$/`, so under any join only the LAST element can match and a `.pem` at `files[0]` rides straight through. `cline-batch-bypass.test.ts` asserts the fan-out catches `sudo` at `commands[1]` and `.env` at `files[1]` with the **real builtins**, and separately asserts the join would have missed them — so a later "simplification" fails loudly rather than going quietly green while enforcement disappears.

Blast radius on shared code is three guards plus an optional 5th parameter on `evaluatePolicies`, not a refactor: the combiner hands its deduped union back through the same per-CLI instruct/allow tails everyone else uses, so none of the 15 existing deny shapes moved, and `expandBatchToolInput` returns `null` for every CLI but cline — the single-shot path is byte-for-byte unchanged for the other 15.

**Two caveats recorded as product decisions, not footnotes:** a deny (`{"cancel":true}`) **aborts the whole run** rather than denying one call, and cline is **fail-open with no opt-out** — the exact opposite of ori, which fails closed for free. `Stop` therefore emits `{"context":…}` instead of a cancel (TaskComplete fires after the task has finished), and the 5 `require-*-before-stop` builtins are inapplicable as on Hermes, Goose and ori.

`apply_patch` is byte-identical to ori's `edit`, so one `splitApplyPatch` now serves both — and it returns per-file old/new text, which reduces ori's multi-file gap to a five-line follow-up. Audit is the thinnest adapter we have: cline already stores **Claude's own content blocks**, with a real `cwd` per session.

**Verified live against cline v3.0.60:** the shipped install path blocked a real two-command `run_commands` batch whose `sudo` sat at index 1 — the run aborted and the harmless first command's side effect never happened — and the verdict named the offending element, `{"cancel":true,"errorMessage":"Blocked Bash [batch 2/2: sudo rm -rf /tmp/zzz]…"}`. Uninstall removed our nine files while preserving a hand-written `TaskStart.py` and leaving the directory in place.

Also bumps to **1.0.4-beta.1** for a beta release, and corrects CLAUDE.md's version-bump instructions: they said to update "only `package.json`", which produces a red CI, because the quality job compares `Cargo.toml`'s workspace version against root `package.json`. A bump moves four files.

## Hermes review

| Field | Value |
|---|---|
| Status | **Approved** |
| Reviewed commit | `01cbadcf0bd195de2d575ffc1842f0419d3b4843` |
| Policy revision | `1d8f31d926828f3bae215c58f5b35baa44acbff0` |
| Model | `gpt-5.6-terra` |
| Duration | 244s |
| Updated | 2026-09-01T12:26:58.529829157+00:00 |

### Summary

Three high-confidence enforcement bypasses remain in the new Grok, Ori, and Cline paths.

### Changes

- Adds Grok, Qwen, Ori, and Cline hook integrations and response handling.
- Adds Cline batch expansion plus Grok payload normalization.
- Adds transcript/audit and daemon collection support for new CLIs.
- Surfaces new CLIs in project dashboard, install CLI, and documentation.

### Validation

None configured.

### Findings

No blocking findings.

<details>
<summary>3 advisory findings</summary>

- **High/High** Route Grok envelopes from Cursor hooks to Grok's contract — The new detector documents that Grok scans the `.cursor` equivalents of Claude hook configuration (normalize-cli-payload.ts:16-19), while Cursor hook entries invoke failproofai with `--cli cursor` (integrations.ts:698-702). However, resolveEffectiveCli only reroutes an identified Grok envelope when the declared CLI is `claude` (normalize-cli-payload.ts:59). A Grok event launched from an installed Cursor hook therefore remains on Cursor normalization and response shaping: its camelCase tool fields are not normalized and Grok receives Cursor's verdict shape, which it does not honor. Tool-specific policies consequently allow these calls. (`src/hooks/normalize-cli-payload.ts:59`)
- **High/High** Do not allow Ori multi-file edits after inspecting only the first path — Ori Edit canonicalization sets `file_path` solely to `paths[0]` (tool-name-canonicalize.ts:162-167). The added regression test explicitly constructs a patch updating `safe.ts` and then `.env`, and asserts that the exposed `file_path` is only `safe.ts` while `.env` is merely retained in `ori_patch_files` (__tests__/hooks/ori-canonicalize.test.ts:115-120). Existing path/content builtins read canonical scalar fields, so block-env-files and block-secrets-write do not see the later patch section; the whole edit is then approved. (`src/hooks/tool-name-canonicalize.ts:167`)
- **High/High** Fail closed rather than collapsing Cline batch tails — Cline fan-out evaluates only the first 32 elements and collapses every remaining item into one synthetic input (batch-fanout.ts:63-69). For Read/Edit, the collapse exposes one `file_path` selected by four secret-name regexes, otherwise the first remaining path (batch-expand.ts:296-302). Thus a 34-item `read_files` call with safe files through index 32 and `/etc/shadow` at index 33 evaluates the collapsed tail as the safe index-32 path. block-read-outside-cwd checks only that scalar target (builtin-policies.ts:2298-2305), allows it, and Cline executes the full batch including the outside-workspace read. (`src/hooks/batch-fanout.ts:63`)

</details>

### Open questions

None.

### Policy overrides

None.

## Summary by CodeRabbit

* **New Features**
  * Added support for Grok, Qwen Code, Cline, and Ori integrations.
  * Added session transcript collection, project discovery, dashboard history, and audit events for the new integrations.
  * Added Cline batch-operation policy enforcement with per-item evaluation.
* **Bug Fixes**
  * Improved CLI detection and payload handling when Grok hooks are routed through another harness.
* **Documentation**
  * Updated supported-harness documentation and quickstart guidance for 16 integrations.
* **Chores**
  * Updated the release version to 1.0.4-beta.1.

### CODE REVIEW

**Comment by hermes-exosphere** [comment]
## Hermes

| | |
|---|---|
| Status | **Reviewing** |
| Verdict | Not reviewed yet |
| Head | `af49df9385ec` |
| Rounds | 0 of 5 |

_No summary yet._

### What this changes

_No component map for this revision._

### Rounds

_No review has finished on this pull request yet._

### Findings

Nothing raised yet.

---

`@hermes-exosphere help` lists every command. This comment is maintained in place — I rewrite it after each review rather than posting a new one.

**Comment by hermes-exosphere** [comment]
## Hermes

| | |
|---|---|
| Status | **Reviewed** |
| Verdict | Approved |
| Head | `01cbadcf0bd1` |
| Rounds | 1 of 5 |

Three high-confidence enforcement bypasses remain in the new Grok, Ori, and Cline paths.

### What this changes

```mermaid
flowchart LR
    n0Hookintegrationregistry["~ Hook integration registry"]
    n1Payloadnormalization["~ Payload normalization"]
    n2Policydecisionengine["~ Policy decision engine"]
    n3Clinebatchenforcement["+ Cline batch enforcement"]
    n4CLIauditadapters["+ CLI audit adapters"]
    n5Daemoncollectors["+ Daemon collectors"]
    n6Projectdashboard["~ Project dashboard"]
    n0Hookintegrationregistry -- "hook payloads" --> n1Payloadnormalization
    n1Payloadnormalization -- "canonical events and inputs" --> n2Policydecisionengine
    n3Clinebatchenforcement -- "per-item evaluations" --> n2Policydecisionengine
    n2Policydecisionengine -- "CLI verdicts" --> n0Hookintegrationregistry
    n4CLIauditadapters -- "session metadata" --> n6Projectdashboard
    n5Daemoncollectors -- "CLI transcript formats" --> n4CLIauditadapters
```

### Rounds

| Round | Reviewed | Commits in this round | Verdict |
|---|---|---|---|
| 1 | `af49df9385ec` | `b58877d7fbe8` `5289b9eb57d3` `f7e489d029a4` `4ded0493d64a` `0b1bda08baaf` `0e48d9111855` `69471ffbe2d5` `9ef8e8f5dd1c` `fccae8f306e4` `ab43d7711c41` `af49df9385ec` | Changes requested — F1 |
| 1 | `01cbadcf0bd1` | `01cbadcf0bd1` | Approved — F4, F5, F6 |

### Findings

**Open**

- **F4** Route Grok envelopes from Cursor hooks to Grok's contract (`src/hooks/normalize-cli-payload.ts`) — noticed at round 2, advisory
- **F5** Do not allow Ori multi-file edits after inspecting only the first path (`src/hooks/tool-name-canonicalize.ts`) — noticed at round 2, advisory
- **F6** Fail closed rather than collapsing Cline batch tails (`src/hooks/batch-fanout.ts`) — noticed at round 2, advisory

**Resolved**

- **F1** Ori treats a failed policy subprocess as an allow (`src/hooks/integrations.ts`) — round 1
- **F2** Ori and Cline session browsers are not wired into the dashboard (`lib/cli-registry.ts`) — round 1
- **F3** README still states that only twelve harnesses are supported (`README.md`) — round 1

---

`@hermes-exosphere help` lists every command. This comment is maintained in place — I rewrite it after each review rather than posting a new one.

File: src/hooks/integrations.ts [hunk 1]
```diff
@@ -2275,13 +2283,727 @@ export const goose: Integration = {
...
+      try { child.kill(); } catch { /* already gone */ }
+      FAIL_OPEN
+        ? finish(done, { permission: "allow" })
+        : finish(fail, new Error("failproofai timed out"));
+    }, 30_000);
+    child.stdout.on("data", (d) => { out += d; });
+    child.on("error", (err) => {
+      FAIL_OPEN ? finish(done, { permission: "allow" }) : finish(fail, err);
+    });
+    child.on("close", () => {
```
Comment (hermes-exosphere) [inline src/hooks/integrations.ts]: **Hermes — High/High (`SEC-001`): Ori treats a failed policy subprocess as an allow**

The generated Ori feature resolves an empty stdout to `{ permission: "allow" }` in `child.on("close")` without inspecting the child exit status. Thus a successfully spawned `node` process that exits non-zero before producing stdout (for example, a missing/corrupt bundled CLI or an evaluator startup failure) returns normally from the provider; `decide()` then maps it to `{ outcome: "allow" }`. Ori's `failureBehavior: "deny"` is never reached on this path.

Required change: Pass the close exit code/signal into the handler and reject on non-zero exit, signal, empty stdout, or any response other than an explicitly validated allow/deny verdict when `FAILPROOFAI_ORI_FAIL_OPEN` is unset. Add a generated-feature test covering a non-zero, empty-output subprocess.

File: lib/cli-registry.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@
 import type { IntegrationType } from "@/src/hooks/types";
 
 /** Canonical CLI ids the registry knows about. Mirrors `INTEGRATION_TYPES`. */
-export const KNOWN_CLI_IDS = ["claude", "codex", "copilot", "cursor", "opencode", "pi", "hermes", "openclaw", "factory", "devin", "antigravity", "goose"] as const satisfies readonly IntegrationType[];
+export const KNOWN_CLI_IDS = ["claude", "codex", "copilot", "cursor", "opencode", "pi", "hermes", "openclaw", "factory", "devin", "antigravity", "goose", "grok", "qwen"] as const satisfies readonly IntegrationType[];
```
Comment (hermes-exosphere) [inline lib/cli-registry.ts]: **Hermes — Medium/High (`COR-001`): Ori and Cline session browsers are not wired into the dashboard**

The PR adds `getOriProjects`/`getClineProjects` and their encoded-name session resolvers, but `getProjectFolders()` imports and merges providers only through Grok and Qwen. The project detail page likewise queries only through Qwen. In addition, `KNOWN_CLI_IDS` ends at Qwen, so Ori and Cline have no filter option or badge metadata. A user with only Ori or Cline sessions therefore sees no project row and cannot open their session history from the dashboard.

Required change: Add Ori and Cline entries to the CLI registry, merge both project providers in `getProjectFolders()`, and query both encoded-name providers in the project page. Extend registry and project-list tests to cover both integrations.

**Review by hermes-exosphere — CHANGES_REQUESTED** [review]
Hermes found blocking issues that should be addressed.

> Review coverage was incomplete, but the concrete blocking findings below are sufficient to request changes.

### High: Ori treats a failed policy subprocess as an allow

- Rule: `SEC-001`
- Location: `src/hooks/integrations.ts:2389`
- Evidence: The generated Ori feature resolves an empty stdout to `{ permission: "allow" }` in `child.on("close")` without inspecting the child exit status. Thus a successfully spawned `node` process that exits non-zero before producing stdout (for example, a missing/corrupt bundled CLI or an evaluator startup failure) returns normally from the provider; `decide()` then maps it to `{ outcome: "allow" }`. Ori's `failureBehavior: "deny"` is never reached on this path.
- Required change: Pass the close exit code/signal into the handler and reject on non-zero exit, signal, empty stdout, or any response other than an explicitly validated allow/deny verdict when `FAILPROOFAI_ORI_FAIL_OPEN` is unset. Add a generated-feature test covering a non-zero, empty-output subprocess.

<details>
<summary>2 advisory findings</summary>

- **Medium/High** Ori and Cline session browsers are not wired into the dashboard — The PR adds `getOriProjects`/`getClineProjects` and their encoded-name session resolvers, but `getProjectFolders()` imports and merges providers only through Grok and Qwen. The project detail page likewise queries only through Qwen. In addition, `KNOWN_CLI_IDS` ends at Qwen, so Ori and Cline have no filter option or badge metadata. A user with only Ori or Cline sessions therefore sees no project row and cannot open their session history from the dashboard. (`lib/cli-registry.ts:30`)
- **Low/High** README still states that only twelve harnesses are supported — The PR adds four new harness logos to the README's grid and changes its documentation link to “All 16”, but the introductory text still says “Twelve harnesses ... ten coding CLIs, plus two gateways.” This is the first support statement readers encounter and now contradicts the shipped integrations and the reference page. (`README.md:33`)

</details>

**Comment by chhhee10** [comment]
Thanks — all three findings were real, and the high one was a genuine hole. Addressed in `01cbadcf`.

### High — Ori treated a failed policy subprocess as an allow (`SEC-001`)

Correct, and worse than a missed edge case: this repo advertises Ori as *the* integration that fails closed for free, and on the one path that matters — failproofai itself being broken — it did the opposite. The claim was written into the changelog, `CLAUDE.md` and this PR's body while the code did the reverse.

Root cause is as described: `child.on("close", () => …)` took no arguments, so a non-zero exit with no stdout resolved to `{permission:"allow"}`. Ori's `failureBehavior:"deny"` could never fire, because **resolving is what tells ori the provider succeeded** — only a rejection reaches that guarantee.

The handler now:

- reads `(code, signal)` and rejects on a signal or a non-zero exit;
- rejects a verdict that is neither an explicit `allow` nor an explicit `deny` — `decide()` maps every non-`"deny"` to allow, so an unvalidated object would silently become one;
- still treats **exit 0 with empty stdout** as an allow, because that is the evaluator's clean-allow shape — rejecting it would deny every permitted tool call on a working install;
- routes every failure branch through `FAILPROOFAI_ORI_FAIL_OPEN`, so the documented escape hatch survives.

Test added as requested: `__tests__/hooks/generated-shim-fail-closed.test.ts`. It asserts on the **shipped artifact** (the install-time generated feature), including that the exit-code guard *precedes* the empty-stdout allow — placed after, it would be dead code. I verified it fails when the bug is reintroduced.

One scoping note, stated plainly: those assertions read the generated source rather than driving a live ori runtime, because the feature imports `ori`, which only exists inside ori's own process. They are a regression guard on the artifact we write to disk. That ori honours the rejection was verified by hand against ori 0.12.0. If you consider that insufficient, say so and I will look at making the verdict logic separately executable.

These generated launchers are strings built at install time, so nothing else typechecks or exercises them — which is exactly how this got in.

### Medium/High — Ori and Cline session browsers not wired into the dashboard

Also correct, and it applied to both new CLIs. The audit adapters worked for `failproofai audit`, but `getProjectFolders()` merged only through Qwen, the project detail page queried only through Qwen, and `KNOWN_CLI_IDS` ended there — so a user with only Ori or Cline sessions saw no project row and could not open their history. Both are now merged, queried, and carry badge metadata.

While fixing it I found a fourth instance not in the review: `lib/download-session.ts` had the same gap. Its exhaustive switch had no arm for either CLI, so a session download would have fallen through the `never` guard. Both are handled now.

The existing unique-badge test then caught a mistake in my own fix — the colours I first chose duplicated existing entries. Now `sky` and `stone`.

### Low/High — README still stated twelve harnesses

Correct. Fixed to sixteen / fourteen coding CLIs.

Worth flagging what made it possible: `copy-counts.test.ts` **pinned the stale wording** rather than deriving it, so the guard passed while the sentence rotted. I deliberately did not change it to derive from `INTEGRATION_TYPES.length` — the purpose of that assertion is that a human re-reads the opening paragraph, and a computed count would stay green while the surrounding prose drifted, which is precisely what happened between the 12th and 16th integrations. The comment now says so.

Four further docs still claimed "all 12 supported harnesses", and `policy-sdk.mdx` claimed "ten of the twelve" for `PostToolUse`; that one is now fourteen of sixteen, counted from `ENFORCEMENT_CAPABILITY` rather than guessed.

---

Gates on the pushed commit: `tsc` clean, **5004 tests passing**, lint 0 errors.

**Review by hermes-exosphere — DISMISSED** [review]
Hermes found no blocking issues in this revision.

<details>
<summary>3 advisory findings</summary>

- **High/High** Route Grok envelopes from Cursor hooks to Grok's contract — The new detector documents that Grok scans the `.cursor` equivalents of Claude hook configuration (normalize-cli-payload.ts:16-19), while Cursor hook entries invoke failproofai with `--cli cursor` (integrations.ts:698-702). However, resolveEffectiveCli only reroutes an identified Grok envelope when the declared CLI is `claude` (normalize-cli-payload.ts:59). A Grok event launched from an installed Cursor hook therefore remains on Cursor normalization and response shaping: its camelCase tool fields are not normalized and Grok receives Cursor's verdict shape, which it does not honor. Tool-specific policies consequently allow these calls. (`src/hooks/normalize-cli-payload.ts:59`)
- **High/High** Do not allow Ori multi-file edits after inspecting only the first path — Ori Edit canonicalization sets `file_path` solely to `paths[0]` (tool-name-canonicalize.ts:162-167). The added regression test explicitly constructs a patch updating `safe.ts` and then `.env`, and asserts that the exposed `file_path` is only `safe.ts` while `.env` is merely retained in `ori_patch_files` (__tests__/hooks/ori-canonicalize.test.ts:115-120). Existing path/content builtins read canonical scalar fields, so block-env-files and block-secrets-write do not see the later patch section; the whole edit is then approved. (`src/hooks/tool-name-canonicalize.ts:167`)
- **High/High** Fail closed rather than collapsing Cline batch tails — Cline fan-out evaluates only the first 32 elements and collapses every remaining item into one synthetic input (batch-fanout.ts:63-69). For Read/Edit, the collapse exposes one `file_path` selected by four secret-name regexes, otherwise the first remaining path (batch-expand.ts:296-302). Thus a 34-item `read_files` call with safe files through index 32 and `/etc/shadow` at index 33 evaluates the collapsed tail as the safe index-32 path. block-read-outside-cwd checks only that scalar target (builtin-policies.ts:2298-2305), allows it, and Cline executes the full batch including the outside-workspace read. (`src/hooks/batch-fanout.ts:63`)

</details>
